From 798fb5aeb16d300904cb81e0965d8077ddcb79e8 Mon Sep 17 00:00:00 2001
From: Rob Hoes <rob.hoes@citrix.com>
Date: Tue, 26 Jul 2022 14:57:42 +0000
Subject: [PATCH 4/5] Total timeout for receiving HTTP headers

Signed-off-by: Rob Hoes <rob.hoes@citrix.com>
(cherry picked from commit 6fd74ea706fd2fd767be691daffe88a3c227814e)
---
 ocaml/xapi/xapi_globs.ml      | 4 ++++
 ocaml/xapi/xapi_mgmt_iface.ml | 1 +
 2 files changed, 5 insertions(+)

diff --git a/ocaml/xapi/xapi_globs.ml b/ocaml/xapi/xapi_globs.ml
index 30d506da2..912ac04b8 100644
--- a/ocaml/xapi/xapi_globs.ml
+++ b/ocaml/xapi/xapi_globs.ml
@@ -880,6 +880,9 @@ let samba_dir = "/var/lib/samba"
 let header_read_timeout_tcp = ref 10.
 (* Timeout in seconds for every read while reading HTTP headers (on TCP only) *)
 
+let header_total_timeout_tcp = ref 60.
+(* Timeout in seconds to receive all HTTP headers (on TCP only) *)
+
 let conn_limit_tcp = ref 800
 
 let conn_limit_unix = ref 1024
@@ -958,6 +961,7 @@ let xapi_globs_spec =
     , Float winbind_update_closest_kdc_interval
     )
   ; ("header_read_timeout_tcp", Float header_read_timeout_tcp)
+  ; ("header_total_timeout_tcp", Float header_total_timeout_tcp)
   ; ("conn_limit_tcp", Int conn_limit_tcp)
   ; ("conn_limit_unix", Int conn_limit_unix)
   ]
diff --git a/ocaml/xapi/xapi_mgmt_iface.ml b/ocaml/xapi/xapi_mgmt_iface.ml
index a12fffebf..89feccc1e 100644
--- a/ocaml/xapi/xapi_mgmt_iface.ml
+++ b/ocaml/xapi/xapi_mgmt_iface.ml
@@ -115,6 +115,7 @@ let start ~__context ?addr () =
   in
   Http_svr.start
     ~header_read_timeout:!Xapi_globs.header_read_timeout_tcp
+    ~header_total_timeout:!Xapi_globs.header_total_timeout_tcp
     ~conn_limit:!Xapi_globs.conn_limit_tcp Xapi_http.server socket ;
   management_interface_server := socket :: !management_interface_server ;
   restart_stunnel ~__context ~accept ;
-- 
2.31.1

